k2/csrc/fsa.h

```diff
@@ -122,6 +126,139 @@ struct Fsa {
+                                // is repeated, so for any valid state 0 <= s <
+                                // num_states we can use arc_indexes[s+1].  That
+                                // is: elements 0 through num_states (inclusive)
+                                // are valid.  CAUTION: arc_indexes[0] may be
+                                // greater than zero.
+
+
+  Arc *arcs;   // Note: arcs[BeginArcIndex()] through arcs[EndArcIndex() - 1]
+               // are valid.
+
+  // Constructor from Fsa
+  Cfsa(const Fsa &fsa);
+
+  Cfsa &operator = (const Cfsa &cfsa) = default;
+  Cfsa(const Cfsa &cfsa) = default;
+
+  Cfsa(int32_t size, int32_t *data);
+
+  int32_t NumStates() const { return num_states; }
+  int32_t FinalState() const { return num_states - 1; }
+};
+
+
+/*
+  Return the number of bytes we'd need to represent this vector of Cfsas
+  linearly as a CfsaVec. */
+size_t GetCfsaVecSize(const std::vector<Cfsa> &fsas_in);
+
+// Return the number of bytes we'd need to represent this Cfsa
+// linearly as a CfsaVec with one element
+size_t GetCfsaVecSize(const Cfsa &fsa_in);
+
+/*
+  Create a CfsaVec from a vector of Cfsas (this involves representing
+  the vector of Fsas in one big linear memory region).
+
+     @param [in] fsas_in  The vector of Cfsas to be linearized;
+                      must be nonempty
+     @param [in] data    The allocated data of size `size` bytes
+     @param [in] size    The size of the memory block passed;
+                         must equal the return value of
+                         GetCfsaVecSize(fsas_in).
+ */
+void CreateCfsaVec(const std::vector<Cfsa> &fsas_in,
+                   void *data,
+                   size_t size);
+
+
+
+class CfsaVec {
+ public:
+
+  /*
+      Constructor from linear data, e.g. in a tensor holding a single
+      Fsa:
+         @param [in] size    size in int32_t elements of `data`, only
+                             needed for checking purposes.
+         @param [in] data    The underlying data.   Format of data is
+                             described below (all elements are of type
+                             int32_t unless stated otherwise).
+
+             - version       Format version number, currently always 1.
+             - num_fsas      The number of FSAs
+             - state_offsets_start  The offset from the start of `data` of
+                             where the `state_offsets` array is, in int32_t
+                             (4-byte) elements.
+             - arc_indexes_start   The offset from the start of `data` of
+                             where the `arc_indexes` array is, in int32_t
+                             (4-byte) elements.
+             - arcs_start    The offset from the start of `data` of where
+                             the first Arc is, in sizeof(Arc) multiples, i.e.
+                             Arc *arcs = ((Arc*)data) + arcs_start
+
+            [possibly some padding here]
```

Comment:
Why would there be some padding here? 

Since all the above fields are `int32_t`, so if the first field is aligned, this array should
also be aligned in memory.

Reply:
Are the paddings for adding extra fields later?

Reply:
On GPUs, I believe data alignment can sometimes be important.  And it might be helpful even on CPU to make sure these things match up with cache lines.